components/script/dom/htmlimageelement.rs

```diff
@@ -366,6 +383,17 @@ impl HTMLImageElement {
         self.upcast::<Node>().dirty(NodeDamage::OtherNodeDamage);
     }
 
+    /// https://html.spec.whatwg.org/multipage/#upgrade-the-pending-request-to-the-current-request
+    fn upgrade_pending_to_current(&self) {
+        let mut pending_opt = self.pending_request.borrow_mut();
+        mem::swap(
+            &mut self.current_request.borrow_mut().deref_mut(),
+            &mut pending_opt.get_or_insert(ImageRequest::new()),
+        );
+        *pending_opt = None;
```

Comment:
I like this suggestion a lot — definitely cleaner, and it is probably a good idea to bail rather than set current request to `ImageRequest::new()` when pending request is `None`.

However, this change has somehow resulted in a regression in some tests.  The behavior _is_ different, but (as far as I can tell) only when `pending_request` is `None`.  I added a `dbg!` to check for `None` in one of the failing tests to see if this was happening in either case.

**Before your suggestion:**
```rust
fn upgrade_pending_to_current(&self) {
    let mut pending_opt = self.pending_request.borrow_mut();
    dbg!(&pending_opt.is_none());
    mem::swap(
        &mut self.current_request.borrow_mut().deref_mut(),
        &mut pending_opt.get_or_insert(ImageRequest::new()),
    );
    *pending_opt = None;
    self.image_request.set(ImageRequestPhase::Current);
}
```

**After your suggestion:**
```rust
fn upgrade_pending_to_current(&self) {
    dbg!(&self.pending_request.borrow().is_none());
    *self.current_request.borrow_mut() = self.pending_request.borrow_mut().take().expect("no pending request");
    self.image_request.set(ImageRequestPhase::Current);
}
```

I ran against [just the parse-a-sizes-attribute-standards-mode.html test](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html), although a few other similar tests failed with similar errors.  I used this command:

```sh
./mach test-wpt --headless /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html
```

In both tests, the `dbg!` output showed `is_none()` was never true and printed the same number of lines, meaning the behavior before and after this change _should_ be the same.  As expected, there were never any crashes logged from the `.expect`, either.

**Output before your suggestion:**

```
0:04.24 INFO Using 8 client processes
 0:04.42 INFO Starting http server on 127.0.0.1:8000
 0:04.42 INFO Starting http server on 127.0.0.1:8001
 0:04.54 INFO Starting https server on 127.0.0.1:8443
 0:05.91 SUITE_START: web-platform-test - running 1 tests
 0:05.92 INFO Starting runner
 0:05.93 TEST_START: /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html
 0:05.93 WARNING Got command init_succeeded in state running
 0:06.77 pid:10971 Full command: /Users/twilco/projects/servo/target/debug/servo --hard-fail -u Servo/wptrunner -Z replace-surrogates -z http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html --user-stylesheet /Users/twilco/projects/servo/resources/ahem.css --certificate-path /Users/twilco/projects/servo/tests/wpt/web-platform-tests/tools/certs/cacert.pem
pid:10971 
 0:06.77 pid:10971 
 0:06.77 pid:10971 
 0:10.07 pid:10971 [components/script/dom/htmlimageelement.rs:398] &pending_opt.is_none() = false
 0:10.07 pid:10971 [components/script/dom/htmlimageelement.rs:398] &pending_opt.is_none() = false
 0:10.07 pid:10971 [components/script/dom/htmlimageelement.rs:398] &pending_opt.is_none() = false
 0:10.07 pid:10971 [components/script/dom/htmlimageelement.rs:398] &pending_opt.is_none() = false
...lots more false dbg statements...
 0:11.14 TEST_END: Test OK. Subtests passed 154/169. Unexpected 0
 0:11.14 INFO No more tests
 0:11.15 INFO Closing logging queue
 0:11.15 INFO queue closed
 0:11.15 INFO Got 0 unexpected results
 0:11.15 SUITE_END
```

**Output after your suggestion:**

```
 0:01.86 INFO Using 8 client processes
 0:01.94 INFO Starting http server on 127.0.0.1:8001
 0:01.94 INFO Starting http server on 127.0.0.1:8000
 0:01.95 INFO Starting https server on 127.0.0.1:8443
 0:02.41 SUITE_START: web-platform-test - running 1 tests
 0:02.43 TEST_START: /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html
 0:02.43 WARNING Got command init_succeeded in state running
 0:02.72 pid:14527 Full command: /Users/twilco/projects/servo/target/debug/servo --hard-fail -u Servo/wptrunner -Z replace-surrogates -z http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html --user-stylesheet /Users/twilco/projects/servo/resources/ahem.css --certificate-path /Users/twilco/projects/servo/tests/wpt/web-platform-tests/tools/certs/cacert.pem
pid:14527 
 0:02.72 pid:14527 
 0:02.72 pid:14527 
 0:05.81 pid:14527 [components/script/dom/htmlimageelement.rs:395] &self.pending_request.borrow().is_none() = false
 0:05.81 pid:14527 [components/script/dom/htmlimageelement.rs:395] &self.pending_request.borrow().is_none() = false
 0:05.81 pid:14527 [components/script/dom/htmlimageelement.rs:395] &self.pending_request.borrow().is_none() = false
...lots more false dbg statements...
 0:06.86 TEST_END: Test OK. Subtests passed 0/169. Unexpected 154
FAIL <img srcset="/images/green-1x1.png?a2 300w, /images/green-16x16.png?a2 301w"> ref sizes="100vw" (standards mode) - assert_unreached: ref currentSrc was "" Reached unreachable code
check/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:13:9
Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
test@http://web-platform.test:8000/resources/testharness.js:544:30
check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:11:5
onload/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:26:5
onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:25:14
FAIL <img srcset="/images/green-1x1.png?b2 450w, /images/green-16x16.png?b2 451w"> ref sizes="100vw" (standards mode) - assert_unreached: ref currentSrc was "" Reached unreachable code
check/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:13:9
Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
test@http://web-platform.test:8000/resources/testharness.js:544:30
check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:11:5
onload/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:26:5
onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:25:14
...lots more failed tests...
 0:06.90 INFO Closing logging queue
 0:06.90 INFO queue closed
```

Perhaps I'm missing something obvious — any idea what is going on here?

Reply:
Wow. I can reproduce this behaviour change, and I'm really confused by it!

Reply:
Update: I've been digging into this. The test is triggering a situation where the img elements start a load, and then a window resize event is received (which is actually intended to be the window's initial size) which kicks off another image load while the first one is in progress. I still have no insight into why the behaviour changes how we write upgrade_pending_to_current.

Reply:
https://github.com/servo/servo/pull/24677 fixes the issues I described in my last comment and makes those `assert_unreached: ref currentSrc was "" Reached unreachable code` test failures go away.

Reply:
That's great — thanks for your continued investigation on this issue!  I'll keep an eye on that PR and rebase after it's landed.

Reply:
So I have a suspicion (based on https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3c8dee3b370e74a01e476e6f9b1152d9) that the reason the behaviour changed is that the original mem::swap code does not actually swap the values. The code that I suggested performs the swap as intended, and this exposed the problems that #24677 fixes.

Reply:
Yep - the presence of the `&mut` before each argument in the mem::swap means that we're swapping `&mut &mut` values. Yikes!

Reply:
I suspect that the 
```rust
    mem::swap(
        &mut self.current_request.borrow_mut().deref_mut(),
        &mut pending_opt.get_or_insert(ImageRequest::new()),
    );
```
was meant to be
```rust
    mem::swap(
        &mut *self.current_request.borrow_mut().deref_mut(),
        &mut *pending_opt.get_or_insert(ImageRequest::new()),
    );
```

Reply:
Well that makes a lot more sense, good catch 😱